File: astro1138/euler31.pl

a(0, _).
a(N, M) :- M > 0, M2 is M - 1, a(N2, M2), N is N2 + 1.

coins([A1, A2, A5, A10, A20, A50, A100, A200]) :-
	a(A1, 200), a(A2, 100), a(A5, 40), a(A10, 20), a(A20, 10), a(A50, 4), a(A100, 2), a(A200, 1),
	A1 * 1 + A2 * 2 + A5 * 5 + A10 * 10 + A20 * 20 + A50 * 50 + A100 * 100 + A200 * 200 is 200.

euler31(NWays) :-
	findall(Coins, coins(Coins), Ways),
	length(Ways, NWays).

File: astro1138/euler31.hs
coins = [(a,b,c,d,e,f,g,h) |
                             a <- [0..200],
                             b <- [0..(100 - a `div` 2)],
                             a + b * 2 <= 200,
                             c <- [0..(40 - (a `div` 5) - (b `div` 2))],
                             d <- [0..20],
                             a + b * 2 + c * 2 + d * 10 <= 200,
                             e <- [0..10],
                             f <- [0..4],
                             g <- [0..2],
                             h <- [0..1],
                             a + b * 2 + c * 2 +
                               d * 10 + e * 20 + f * 50 +
                                 g * 100 + h * 200 == 200]

euler_31 = length coins

main = do
  putStrLn $ show $ euler_31
